ocaml/xapi/vm_sysprep.ml

```diff
@@ -0,0 +1,267 @@
+    )
+  with e ->
+    let msg = Printexc.to_string e in
+    Helpers.internal_error "%s failed: %s" __FUNCTION__ msg
+
+(** create a local ISO SR when necessary and update it such that it
+    recognises any ISO we added or removed *)
+let update_sr ~__context =
+  let host = Helpers.get_localhost ~__context in
+  let hostname = Db.Host.get_hostname ~__context ~self:host in
+  let label = SR.name hostname in
+  let mib n = Int64.(n * 1024 * 1024 |> of_int) in
+  let sr =
+    match SR.find_opt ~__context ~label with
+    | Some sr ->
+        sr
+    | None ->
+        let device_config = [("location", SR.dir); ("legacy_mode", "true")] in
+        call ~__context @@ fun rpc session_id ->
+        Client.SR.create ~rpc ~session_id ~host ~name_label:label ~device_config
+          ~content_type:"iso" ~_type:"iso" ~name_description:"Sysprep ISOs"
+          ~shared:false ~sm_config:[] ~physical_size:(mib 512)
+  in
+  call ~__context @@ fun rpc session_id ->
+  Client.SR.scan ~rpc ~session_id ~sr ;
+  sr
+
+(** Find the VBD for the CD drive on [vm] *)
+let find_cdr_vbd ~__context ~vm =
+  let vbds = Db.VM.get_VBDs ~__context ~self:vm in
+  let vbds' =
+    List.map (fun self -> (self, Db.VBD.get_record ~__context ~self)) vbds
+  in
+  let is_cd (_rf, rc) =
+    let open API in
+    rc.vBD_type = `CD && rc.vBD_empty
+  in
+  let uuid = Db.VM.get_uuid ~__context ~self:vm in
+  match List.filter is_cd vbds' with
+  | [] ->
+      fail VM_CDR_not_found
+  | [(rf, rc)] ->
+      debug "%s: for VM %s using VBD %s" __FUNCTION__ uuid rc.API.vBD_uuid ;
+      rf
+  | (rf, rc) :: _ ->
+      debug "%s: for VM %s using VBD %s" __FUNCTION__ uuid rc.API.vBD_uuid ;
+      warn "%s: for VM %s found additions VBDs" __FUNCTION__ uuid ;
+      rf
+
+(** Find the VDI that contains the unattend.xml based on its name. This
+    should be unique *)
+let find_vdi ~__context ~label =
+  match Db.VDI.get_by_name_label ~__context ~label with
+  | [] ->
+      internal_error "%s: can't find VDI for %s" __FUNCTION__ label
+  | [vdi] ->
+      vdi
+  | vdi :: _ ->
+      warn "%s: more than one VDI with label %s" __FUNCTION__ label ;
+      vdi
+
+(** notify the VM with [domid] to run sysprep and where to find the
+    file. *)
+let trigger ~domid =
+  let open Ezxenstore_core.Xenstore in
+  let control = Printf.sprintf "/local/domain/%Ld/control/sysprep" domid in
+  with_xs (fun xs ->
+      xs.Xs.write (control // "filename") "D://unattend.xml" ;
+      xs.Xs.write (control // "action") "sysprep" ;
+      debug "%s: notified domain %Ld" __FUNCTION__ domid ;
+      let rec wait n =
+        match (n, xs.Xs.read (control // "action")) with
+        | _, "running" ->
+            "running"
+        | n, action when n < 0 ->
+            action
+        | _, _ ->
+            Thread.delay 1.0 ;
+            wait (n - 1)
+      in
+      (* wait up to 5 iterations for runnung to appear or report whatever
+         is the status at the end *)
+      wait 5
```

Comment:
We have some code already that waits for a xenstore key to reach a certain value or timeout, using xenstore watches.

See xc/device.ml for examples, I think it is `Watch.value_to_become` from ezxenstore.

That might then also avoid a race condition where the VM reboots in less than a second and causes this loop to fail (the watch code has a race condition too, but the window is much smaller: it gets woken immediately when it processes the watch event and then reads the xenstore key).